File: Chapter-3/Chapter-3.scala
//Code for Chapter 3

// Example 3-2
val lines = sc.parallelize(List("pandas", "i like pandas"))



// Example 3-4
val inputRDD = sc.textFile("log.txt")
val errorsRDD = inputRDD.filter(line => line.contains("error"))


// Example 3-7
println("Input had " + badLinesRDD.count() + " concerning lines")
println("Here are 10 examples:")
badLinesRDD.take(10).foreach(println)



// Example 3-9
val input = sc.parallelize(List(1, 2, 3, 4))
val result = input.map(x => x * x)
println(result.collect().mkString(","))



// Example 3-11
val sum = rdd.reduce((x, y) => x + y)


// Example 3-13
val result = input.aggregate((0, 0))(
                               (acc, value) => (acc._1 + value, acc._2 + 1),
                               (acc1, acc2) => (acc1._1 + acc2._1, acc1._2 + acc2._2))
val avg = result._1 / result._2.toDouble



// Example 3-17
class SearchFunctions(val query: String) {
     def isMatch(s: String): Boolean = {
          s.contains(query)
     }
     def getMatchesFunctionReference(rdd: RDD[String]): RDD[String] = {
          // Problem: "isMatch" means "this.isMatch", so we pass all of "this"
          rdd.map(isMatch)
     }
     def getMatchesFieldReference(rdd: RDD[String]): RDD[String] = {
          // Problem: "query" means "this.query", so we pass all of "this"
          rdd.map(x => x.split(query))
     }
     def getMatchesNoReference(rdd: RDD[String]): RDD[String] = {
          // Safe: extract just the field we need into a local variable
          val query_ = this.query
          rdd.map(x => x.split(query_))
     }
}



// Example 3-18
val result = input.map(x => x * x)
result.persist(StorageLevel.DISK_ONLY)
println(result.count())
println(result.collect().mkString(","))

File: Chapter-3/Chapter-3.py
# Code for Chapter 3
# as of Python 2.x


# Example 3-1
 lines = sc.parallelize(["pandas", "i like pandas"])

  
# Example 3-2
inputRDD = sc.textFile("log.txt")
errorsRDD = inputRDD.filter(lambda x: "error" in x)



# Example 3-5
errorsRDD = inputRDD.filter(lambda x: "error" in x)
warningsRDD = inputRDD.filter(lambda x: "warning" in x)
badLinesRDD = errorsRDD.union(warningsRDD)



# Example 3-6
print "Input had " + badLinesRDD.count() + " concerning lines"
print "Here are 10 examples:"
for line in badLinesRDD.take(10):
       print line
        
        
        
# Example 3-8
nums = sc.parallelize([1, 2, 3, 4])
squared = nums.map(lambda x: x * x).collect()
for num in squared:
      print "%i " % (num)
        
        
# Example 3-10
sum = rdd.reduce(lambda x, y: x + y)        
        
    
        
# Example 3-12
sumCount = nums.aggregate((0, 0),
                               (lambda acc, value: (acc[0] + value, acc[1] + 1)), 
                               (lambda acc1, acc2: (acc1[0] + acc2[0], acc1[1] + acc2[1])))
avg = sumCount[0] / float(sumCount[1])



# Example 3-14
word = rdd.filter(lambda s: "error" in s)
def containsError(s):
       return "error" in s
word = rdd.filter(containsError)



# Example 3-15
class SearchFunctions(object):
      def __init__(self, query):
             self.query = query
      def isMatch(self, s):
             return self.query in s
      def getMatchesFunctionReference(self, rdd):
             # Problem: references all of "self" in "self.isMatch"		
             return rdd.filter(self.isMatch)
      def getMatchesMemberReference(self, rdd):
             # Problem: references all of "self" in "self.query"
             return rdd.filter(lambda x: self.query in x)


    
    # Example 3-16
    class WordFunctions(object):
      ...
      def getMatchesNoReference(self, rdd):
            # Safe: extract only the field we need into a local variable
            query = self.query
            return rdd.filter(lambda x: query in x)

